vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/QueryImpl.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright (C) 2017 Julien Viet
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.vertx.sqlclient.impl;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Promise;
+import io.vertx.sqlclient.Query;
+import io.vertx.sqlclient.Row;
+import io.vertx.sqlclient.RowSet;
+import io.vertx.sqlclient.SqlResult;
+import io.vertx.sqlclient.Tuple;
+import io.vertx.sqlclient.impl.command.BiCommand;
+import io.vertx.sqlclient.impl.command.PrepareStatementCommand;
+
+import java.util.List;
+
+/**
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+public class QueryImpl<T, R extends SqlResult<T>> extends QueryBase<T, R> {
+
+  static Query<RowSet<Row>> create(SqlClientBase<?> client, boolean autoCommit, boolean singleton, boolean prepared, String sql) {
+    SqlResultBuilder<RowSet<Row>, RowSetImpl<Row>, RowSet<Row>> builder = new SqlResultBuilder<>(RowSetImpl.FACTORY, RowSetImpl.COLLECTOR);
+    return new QueryImpl<>(client, autoCommit, singleton, prepared, sql, builder);
+  }
+
+  private final SqlClientBase<?> scheduler;
+  private final boolean singleton;
+  private final boolean prepared;
+  private final String sql;
+
+  public QueryImpl(SqlClientBase<?> client, boolean autoCommit, boolean singleton, boolean prepared, String sql, SqlResultBuilder<T, ?, R> builder) {
+    super(autoCommit, builder);
+    this.scheduler = client;
+    this.singleton = singleton;
+    this.prepared = prepared;
+    this.sql = sql;
+  }
+
+  @Override
+  protected <T2, R2 extends SqlResult<T2>> QueryBase<T2, R2> copy(SqlResultBuilder<T2, ?, R2> builder) {
+    return new QueryImpl<>(scheduler, autoCommit, singleton, prepared, sql, builder);
+  }
+
+  private void execute(Tuple arguments, Promise<R> promise) {
+    SqlResultHandler handler = builder.createHandler(promise);
+    if (prepared) {
+      BiCommand<PreparedStatement, Boolean> abc = new BiCommand<>(new PrepareStatementCommand(sql), ps -> {
+        String msg = ps.prepare((TupleInternal) arguments);
+        if (msg != null) {
+          return Future.failedFuture(msg);
+        }
+        return Future.succeededFuture(builder.createCommand(ps, autoCommit, arguments, handler));
+      });
+      scheduler.schedule(abc, handler);
+    } else {
+      builder.execute(scheduler, sql, autoCommit, singleton, handler);
```

Comment:
I'm afraid users would be confused here especially those who have used client side prepared statement before, if someone writes code like this
```
conn.createQuery("SELECT * FROM users WHERE id = ?")
        .execute(Tuple.of(1));
```

They will get wrong sql syntax error but what they expect is executing a query with a param.

And even this kind of query can also be executed
```
conn.createQuery("SELECT * FROM users")
        .execute(Tuple.of(1, 2, 3));
```

Reply:
it is not cleat what you mean, you are commenting on `QueryImpl` which is not the API

Reply:
I mean the API allows that a simple query could be executed or batched with parameters but it's implicitly executed without any parameters

Reply:
Can you clarify your concern here @BillyYccc ?

If a user just does `conn.createQuery("SELECT * FROM users WHERE id = ?")` no interaction with the DB would occur until `execute()` is invoked. So the first code example you showed will work fine.

For the second code example, users can already use incorrect SQL and params with the current API like this:
```java
// blows up because SQL has 0 params but 3 were passed in
conn.preparedQuery("SELECT * FROM users", Tuple.of(1, 2, 3));
```

Reply:
I think they are different in terms of a user's intuitions.

My first example shows a scene where users wants to execute a client side prepared statement which is commonly used in JDBC drivers, but we don't support this in all clients so users will get an error of wrong sql syntax, this looks strange for those users but it's OK once users know we don't support this kind of queries.

My second example shows a usecase when users accidently writes a sql of no placeholders but execute with several parameters, the query could be executed and parameters are ignored but in my opinion using simple query with parameters should be forbidden rather than executing the simple query implicitly while ignoring params if we don't support client side prepared statement. 

As for the prepared statement example, it's different because the prepared statement has been pre-compiled, when you execute the prepared statement, the params count and even types will be validated with param metadata locally so users will quickly get an error signal without executing it in reality.

Reply:
Ah, I think I see what you are getting at. For the first example, it is a syntax error because the user is doing `con.createQuery(SQL)` as opposed to `con.createPreparedQuery()`?
This does indeed seem like an issue. Since `Query` is used for both prepared and non-prepared queries.